File: options/typess/portfolio.py
import numpy as np
import pandas as pd

from datetime import date
from collections.abc import Mapping
from typing import Dict, Set, List

from google.protobuf.internal.containers import MessageMap
from options.typess.cash import Cash
from options.typess.equity import Equity
from options.typess.holding import Holding
from options.typess.option import Option
from options.typess.security import Security, SecurityDataSnap
from options.typess.scenario import Scenario
from shared.modules.logger import warning


class Portfolio(Mapping):
    holdings: Dict[Security, float]

    def __init__(self, holdings: Dict[Security, float] = None):
        self.holdings = holdings or {}

    @classmethod
    def from_message_map(cls, msg_map: MessageMap[str, object], calculation_date: date = None):
        return cls.from_holdings({Holding.from_holding_pb(v, calculation_date) for v in msg_map.values()})

    @classmethod
    def from_holdings(cls, holdings: Set[Holding]):
        return cls({h.symbol: h.quantity for h in holdings})

    @property
    def underlying(self) -> str:
        return next(iter(self.holdings.keys())).underlying_symbol.upper() if len(self.holdings) > 0 else None

    def add_holding(self, security: Security, quantity: float):
        if security not in self.holdings:
            self.holdings[security] = 0
        self.holdings[security] += quantity

    def remove_security(self, security):
        self.holdings.pop(security) if security in self.holdings else None

    def delta_total(self, calc_date: date, s: float, df: pd.DataFrame, iv_col: str = 'mid_iv') -> float:
        from options.helper import val_from_df
        _delta_total = 0
        for sec, q in self.holdings.items():
            if isinstance(sec, Equity):
                _delta_total += q
            elif isinstance(sec, Option):
                try:
                    iv = val_from_df(df, sec.expiry, sec.optionContract.strike, sec.right, iv_col)
                    if np.isnan(iv) or iv == 0:
                        price = val_from_df(df, sec.expiry, sec.optionContract.strike, sec.right, 'mid_price')
                        spot = val_from_df(df, sec.expiry, sec.optionContract.strike, sec.right, 'spot')
                        iv = sec.iv(price, spot, calc_date)
                except (IndexError, KeyError):
                    # Missing price data. Illiquid. Can fill with model IV, interpolated IV.
                    warning(f'No {iv_col} for {sec} on {calc_date}. Setting IV to 0.2. To be improved. Some model IV better in this case')
                    iv = 0.2
                _delta_total += sec.delta(iv, s, calc_date) * q * sec.multiplier
            elif isinstance(sec, Cash):
                continue
            else:
                raise ValueError(f'Invalid security type: {type(sec)}')
        return _delta_total

    def print_backtest_holdings(self):
        print('Symbol,Quantity,FillPrice')
        for sec, q in self.holdings.items():
            print(f'{str(sec)},{q},0')

    def get_holdings(self) -> List[Holding]:
        return [Holding(sec, q) for sec, q in self.holdings.items()]

    def nlv(self, market_data: Dict[Security, SecurityDataSnap], scenario: Scenario | str = Scenario.mid):
        return sum((security.nlv(market_data, quantity, scenario) for security, quantity in self.holdings.items()))
        # {(security, quantity): security.nlv(market_data, quantity, scenario) for security, quantity in self.holdings.items()}
        # sum({(security, quantity): security.nlv(market_data, quantity, scenario) for security, quantity in self.holdings.items()}.values())

    def keys(self):
        return self.holdings.keys()

    def securities(self):
        return self.holdings.keys()

    def __iter__(self):
        return iter(self.get_holdings())

    def __getitem__(self, __key):
        return self.holdings[__key] if __key in self.holdings else 0

    def __len__(self):
        return len(self.holdings)

    def __repr__(self):
        return ', '.join([f'{str(o)}: {q}' for o, q in self.holdings.items()])

    def __copy__(self):
        return Portfolio(self.holdings)

    def __bool__(self):
        return bool(self.holdings)


if __name__ == '__main__':
    pf = Portfolio({Equity('AAPL'): 100})
    for h in pf:
        print(h)

File: fino/src/types/portfolio.jl
# types/Portfolio.jl

using DataFrames
using Dates
using .Fino: Holding

"""
    Portfolio

A collection of security holdings (implements dict-like interface).

Fields:
- `holdings::Dict{Security, Float64}` — security => quantity mapping
"""
mutable struct Portfolio
    holdings::Dict{Security, Float64}

    function Portfolio(holdings::Union{Dict{Security, Float64}, Nothing} = nothing)
        new(holdings === nothing ? Dict{Security, Float64}() : holdings)
    end
end

# ── Factory methods ────────────────────────────────────────────────────────────

"""
    Portfolio(holdings::Set{Holding})

Create a Portfolio from a set of Holdings.
"""
function Portfolio(holdings::Vector{Holding})
    pf_dict = Dict{Security, Float64}()
    for h in holdings
        pf_dict[h.symbol] = h.quantity
    end
    return Portfolio(pf_dict)
end

# ── Properties ─────────────────────────────────────────────────────────────────

"""
    underlying(pf::Portfolio)::Union{String, Nothing}

Return the underlying symbol if portfolio is non-empty.
"""
function underlying(pf::Portfolio)::Union{String, Nothing}
    if length(pf.holdings) > 0
        first_security = first(keys(pf.holdings))
        return uppercase(underlying_symbol(first_security))
    end
    return nothing
end

# ── Portfolio operations ───────────────────────────────────────────────────────

"""
    add_holding!(pf::Portfolio, security::Security, quantity::Float64)

Add quantity to a security in the portfolio.
"""
function add_holding!(pf::Portfolio, security::Security, quantity::Float64)::Portfolio
    if !haskey(pf.holdings, security)
        pf.holdings[security] = 0.0
    end
    pf.holdings[security] += quantity
    return pf
end

"""
    remove_security!(pf::Portfolio, security::Security)

Remove a security from the portfolio.
"""
function remove_security!(pf::Portfolio, security::Security)
    delete!(pf.holdings, security)
    return pf
end

"""
    delta_total(pf::Portfolio, calc_date::Date, s::Float64, df::DataFrame; iv_col::String = "mid_iv")::Float64

Calculate total delta of portfolio across all positions.
"""
function delta_total(pf::Portfolio, calc_date::Date, s::Float64, df::DataFrame; iv_col::String = "mid_iv")::Float64
    _delta_total = 0.0

    for (sec, q) in pf.holdings
        if typeof(sec) <: Equity
            _delta_total += q
        elseif typeof(sec) <: Option
            try
                # val_from_df(df, sec.expiry, sec.strike, sec.right, iv_col)
                # Assuming val_from_df is a helper function
                iv_val = val_from_df(df, sec.expiry, sec.optionContract.strike, sec.right, iv_col)

                if isnan(iv_val) || iv_val == 0.0
                    price = val_from_df(df, sec.expiry, sec.optionContract.strike, sec.right, "mid_price")
                    spot = val_from_df(df, sec.expiry, sec.optionContract.strike, sec.right, "spot")
                    iv_val = iv(sec, price, spot, calc_date)
                end
            catch e
                # Missing price data
                @warn "No $iv_col for $sec on $calc_date. Setting IV to 0.2. To be improved."
                iv_val = 0.2
            end
            _delta_total += delta(sec, iv_val, s, calc_date) * q * sec.multiplier
        elseif typeof(sec) <: Cash
            continue
        else
            throw(ArgumentError("Invalid security type: $(typeof(sec))"))
        end
    end

    return _delta_total
end

"""
    get_holdings(pf::Portfolio)::Vector{Holding}

Return all holdings as a vector of Holding objects.
"""
function get_holdings(pf::Portfolio)::Vector{Holding}
    return [Holding(sec, q) for (sec, q) in pf.holdings]
end

"""
    nlv(pf::Portfolio, market_data::Dict{Security, SecurityDataSnap}, scenario::String = "mid")::Float64

Calculate net liquidation value of portfolio.
"""
function nlv(pf::Portfolio, market_data::Dict{Security, SecurityDataSnap}, scenario::String = "mid")::Float64
    return sum(nlv(security, market_data, quantity, scenario) for (security, quantity) in pf.holdings)
end

"""
    print_backtest_holdings(pf::Portfolio)

Print holdings in CSV format (Symbol, Quantity, FillPrice).
"""
function print_backtest_holdings(pf::Portfolio)
    println("Symbol,Quantity,FillPrice")
    for (sec, q) in pf.holdings
        println("$(string(sec)),$q,0")
    end
end

# ── Dict-like interface ────────────────────────────────────────────────────────

    Base.keys(pf::Portfolio) = keys(pf.holdings)

function securities(pf::Portfolio)
    return keys(pf.holdings)
end

    Base.iterate(pf::Portfolio) = iterate(get_holdings(pf))
    Base.iterate(pf::Portfolio, state) = iterate(get_holdings(pf), state)

    Base.getindex(pf::Portfolio, key::Security) = get(pf.holdings, key, 0.0)

    Base.get(pf::Portfolio, key::Security, default) = get(pf.holdings, key, default)

    Base.length(pf::Portfolio) = length(pf.holdings)

    Base.repr(pf::Portfolio) = join(["$(string(o)): $q" for (o, q) in pf.holdings], ", ")

    Base.string(pf::Portfolio) = repr(pf)

"""
    copy(pf::Portfolio)

Create a shallow copy of the portfolio.
"""
    Base.copy(pf::Portfolio) = Portfolio(copy(pf.holdings))

    Base.isempty(pf::Portfolio) = isempty(pf.holdings)

# ── Serialization ──────────────────────────────────────────────────────────────

"""
    getstate(pf::Portfolio)

Return a serializable dict representation.
"""
function getstate(pf::Portfolio)::Dict{String, Any}
    return Dict("holdings" => pf.holdings)
end

"""
    setstate!(pf::Portfolio, state::Dict)

Restore portfolio from serialized state.
"""
function setstate!(pf::Portfolio, state::Dict{String, Any})
    pf.holdings = state["holdings"]
    return pf
end
